### ISSUE TITLE
Lock files with fcntl while syncing (for syncing SQLite databases, etc)

### ISSUE THREAD

**Issue by argv-minus-one** [syncthing/syncthing#4242]
While syncing files on POSIX platforms, please consider acquiring locks on those files with `fcntl`, to help avoid file corruption.

My use case is syncing Firefox profiles. Currently, this isn't safe, because Firefox uses SQLite databases that are likely to be written to in the middle of syncing, potentially corrupting them. (I found this out the hard way, when my password database got corrupted. Good thing Syncthing keeps old versions!) But [SQLite uses `fcntl`-based file locking on POSIX platforms](https://sqlite.org/faq.html#q5), so if Syncthing did too, then it should be safe.

**Comment by AudriusButkevicius** [comment]
Syncthing is not for syncing databases.

**Comment by calmh** [comment]
This could help from the perspective that if we are able to acquire the lock it means the database is not open, so it is consistent with itself and any WAL file etc. Apart from that, we already handle (in a way) the fact that the file might change during syncing - the data blocks will have unexpected hashes and be rejected on the receiving side.

But using locks has the downside that Firefox etc would fail to open the file if it wanted to while we were hashing it, making Syncthing obtrusive and annoying.

**Comment by argv-minus-one** [comment]
@AudriusButkevicius Considering how many applications use SQLite to store their data, that seems like a rather large caveat. Is the average user really going to know that some app's save or settings file is actually a database and therefore unsafe to sync? I certainly didn't!

How many more people's files have to get unexpectedly corrupted before you take this issue seriously?

@calmh It wouldn't fail to open. It would wait for the lock to be released, only if Firefox and/or Syncthing is trying to *write* to the database. I can see that getting annoying, but data corruption is still worse.

**Comment by calmh** [comment]
Holding a read (shared) lock would prevent Firefox (or anyone else) from taking a write (exclusive) lock, which it presumably needs to be able to use the profile.

Note that the corruption here is mostly self correcting as well. If the file changes while we hash it, the subsequent sync will fail due to hash mismatches and won’t happen. The corruption would occur because we managed to hash and sync a “snapshot” of the file that was quiescent on disk but still invalid. Presumably it would get flushed / synced reasonable soon after such a state and next sync would bring it up to speed.

But no, file locking is a mess and doesn’t really give us any guarantees anyway. SQLite may be fairly common, but we’re not going to implement a special case for it nonetheless.

**Comment by AudriusButkevicius** [comment]
We do what we are asked, sync files, if you happened to use it on an application that can't deal with that, there is not an awful lot we can do.

What if the user syncs a database that doesn't use flocks? Are we the ones to blame as well?

I think this is a battle we can't win.

**Comment by calmh** [comment]
Oh, and if you are thinking of doing *bidirectional* syncing of SQLite files (i.e., open on both sides and synced in both directions), this will break spectacularly not matter what sort of locking we implement. Do not do that. We cannot make it work.

**Comment by argv-minus-one** [comment]
>Holding a read (shared) lock would prevent Firefox (or anyone else) from taking a write (exclusive) lock, which it presumably needs to be able to use the profile.

[SQLite normally only holds locks for a few milliseconds at a time.](https://sqlite.org/faq.html#q5)

>Note that the corruption here is mostly self correcting as well. If the file changes while we hash it, the subsequent sync will fail due to hash mismatches and won’t happen.

After the incident, I looked through Syncthing's logs and found that, yes, it was mostly detecting that situation and backing off correctly. But it obviously isn't guaranteed.

>But no, file locking is a mess, and doesn’t really give us any guarantees anyway.

You mean because `flock`, `fcntl`, and `lockf` locks are separate? That *is* a mess. Three different lock styles, which may or may not interact with each other. If you try to acquire all three locks at the same time, you risk deadlocking. You could acquire only an `fcntl` lock and hope for the best, but as you say, no guarantees. Linux supports mandatory locking, but it doesn't work correctly. *sigh*

For what it's worth, on the Linux machines I have access to, most of the locks held at the moment are of the POSIX (`fcntl` or `lockf`) type. So, using that kind of lock would at least decrease the probability of corruption.

>SQLite may be fairly common, but we’re not going to implement a special case for it nonetheless.

Heh. I was originally going to ask for special handling for SQLite databases (detect that a file is a SQLite database, and lock it using the same protocol as SQLite itself, i.e. `fcntl`). I guess that's my answer.

>Oh, and if you are thinking of doing \*bidirectional\* syncing of SQLite files (i.e., open on both sides and synced in both directions), this will break spectacularly not matter what sort of locking we implement. Do not do that. We cannot make it work.

That's what I was doing when I got the corruption. It did indeed break spectacularly.

You say it won't work even *with* locking? Why is that?

**Comment by AudriusButkevicius** [comment]
http://www.sqlite.org/faq.html#q5

```
But only one process can be making changes to the database at any moment in time, however.
```

**Comment by argv-minus-one** [comment]
That means other processes will *wait*, not throw errors.

SQLite is specifically designed to allow multiple processes to safely use the same database at the same time, by acquiring `fcntl` locks while reading/writing.

**Comment by AudriusButkevicius** [comment]
I know for sure that it doesn't work on multiple threads if you don't compile our own version with magical switches, so I would never trust it to work with multiple processes.

Their FAQ clearly states that two processes should not be changing the database at any moment in time.

We can argue until the cows go home, I don't think we should invest into special cases ever, because soon after this example we will be expected to create .~xls files to lock office files and what not.

If you are interested in having distributed sqlite, see: https://github.com/rqlite/rqlite
If you are interested in syncing your firefox profile see https://www.mozilla.org/en-US/firefox/sync/

Both of the problems are solved, I am not sure why we are still discussing this.

Also, your claim seems false atleast according to stack overflow.
https://stackoverflow.com/questions/25319031/what-is-the-default-value-of-sqlite3-busy-timeout

By default, there is no timeout and it throws errors immediately. Best case it throws errors after the specified timeout.

**Comment by argv-minus-one** [comment]
>If you are interested in syncing your firefox profile see https://www.mozilla.org/en-US/firefox/sync/

[Firefox Sync apparently also has data corruption issues.](https://support.mozilla.org/en-US/questions/1072741)

>Also, your claim seems false atleast according to stack overflow.
>https://stackoverflow.com/questions/25319031/what-is-the-default-value-of-sqlite3-busy-timeout

>By default, there is no timeout and it throws errors immediately. Best case it throws errors after the specified timeout.

I see. Then syncing SQLite databases safely is impossible, even with locking, so I give up.

**Comment by jult** [comment]
I too wanted to sync browser profiles, only to find this caused all kinds of issues. From add-ons being auto-updated on one instance but not on the other, etc. I stepped away from it, the browser do a great job by themselves, as the latest posts above mine explain. But it can be handy to sync bookmark and other static files within profiles, which does work fine with syncthing. After some time I gathered a global bunch of ignores for use with SyncThing, which helped a lot, also against the dreaded conflicts; https://gist.github.com/jult/e2eaedad6b9e29d95977fea0ddffae7d
